src/persist-client/src/internal/state_fetcher.rs

```diff
@@ -0,0 +1,373 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A helper for fetching incrementally updated state at various versions.
+
+use std::fmt::Debug;
+use std::marker::PhantomData;
+use std::ops::ControlFlow::{Break, Continue};
+use std::sync::Arc;
+use std::time::SystemTime;
+
+use bytes::Bytes;
+use differential_dataflow::difference::Semigroup;
+use differential_dataflow::lattice::Lattice;
+use mz_persist::location::{Atomicity, Blob, Consensus, SeqNo, VersionedData};
+use mz_persist::retry::Retry;
+use mz_persist_types::{Codec, Codec64};
+use timely::progress::Timestamp;
+use tracing::{debug, trace};
+
+use crate::error::CodecMismatch;
+use crate::internal::machine::retry_external;
+use crate::internal::paths::{PartialRollupKey, RollupId};
+use crate::internal::state::State;
+use crate::internal::state_diff::StateDiff;
+use crate::{Metrics, PersistConfig, ShardId};
+
+/// A helper for fetching incrementally updated state at various versions.
+///
+/// WIP Invariants:
+/// - diffs for current.seqno_since..=current exist
+/// - all live seqnos are constructable
+///   - it follows that there must be a rollup at the smallest live seqno or the
+///     smallest live seqno-1. we enforce the former
+/// - rollups for seqnos < smallest live seqno are leaked
+///   - this means not all rollups referenced by live seqnos exist
```

Comment:
hmm not sure I follow how the statement implies the second (also leaked isn't defined, so not quiiiiite sure I have the same meaning in my head)

Reply:
this is all precisely defined now